server/datastore/mysql/software.go

```diff
@@ -76,16 +80,152 @@ func (d *Datastore) SaveHostSoftware(host *fleet.Host) error {
 	return nil
 }
 
-func (d *Datastore) LoadHostSoftware(host *fleet.Host) error {
-	host.HostSoftware = fleet.HostSoftware{Modified: false}
+func nothingChanged(current []fleet.Software, incoming []fleet.Software) bool {
+	if len(current) != len(incoming) {
+		return false
+	}
+
+	currentBitmap := make(map[string]bool)
+	for _, s := range current {
+		currentBitmap[softwareToUniqueString(s)] = false
+	}
+	for _, s := range incoming {
+		if _, ok := currentBitmap[softwareToUniqueString(s)]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (d *Datastore) applyChangesForNewSoftware(tx *sqlx.Tx, host *fleet.Host) error {
+	storedCurrentSoftware, err := d.hostSoftwareFromHostID(tx, host.ID)
+	if err != nil {
+		return errors.Wrap(err, "loading current software for host")
+	}
+
+	if nothingChanged(storedCurrentSoftware, host.Software) {
+		return nil
+	}
+
+	current := softwareSliceToIdMap(storedCurrentSoftware)
+	incoming := softwareSliceToSet(host.Software)
+
+	if err = d.deleteUninstalledHostSoftware(tx, host.ID, current, incoming); err != nil {
+		return err
+	}
+
+	if err = d.insertNewInstalledHostSoftware(tx, host.ID, current, incoming); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *Datastore) deleteUninstalledHostSoftware(
+	tx *sqlx.Tx,
+	hostID uint,
+	currentIdmap map[string]uint,
+	incomingBitmap map[string]bool,
+) error {
+	var deletesHostSoftware []interface{}
+	deletesHostSoftware = append(deletesHostSoftware, hostID)
+
+	for currentKey := range currentIdmap {
+		if _, ok := incomingBitmap[currentKey]; !ok {
+			deletesHostSoftware = append(deletesHostSoftware, currentIdmap[currentKey])
+			// TODO: delete from software if no host has it
+		}
+	}
+	if len(deletesHostSoftware) > 1 {
+		sql := fmt.Sprintf(
+			`DELETE FROM host_software WHERE host_id = ? AND software_id IN (%s)`,
+			strings.TrimSuffix(strings.Repeat("?,", len(deletesHostSoftware)-1), ","),
+		)
+		if _, err := tx.Exec(sql, deletesHostSoftware...); err != nil {
+			return errors.Wrap(err, "delete host software")
+		}
+	}
+	return nil
+}
+
+func (d *Datastore) getOrGenerateSoftwareId(tx *sqlx.Tx, s fleet.Software) (uint, error) {
+	var existingId []int64
+	if err := tx.Select(
+		&existingId,
+		`SELECT id FROM software WHERE name = ? and version = ? and source = ?`,
+		s.Name, s.Version, s.Source,
+	); err != nil {
+		return 0, err
+	}
+	if len(existingId) > 0 {
+		return uint(existingId[0]), nil
+	}
+
+	result, err := tx.Exec(
+		`INSERT IGNORE INTO software (name, version, source) VALUES (?, ?, ?)`,
+		s.Name, s.Version, s.Source,
+	)
+	if err != nil {
+		return 0, errors.Wrap(err, "insert software")
+	}
+	id, err := result.LastInsertId()
+	if err != nil {
+		return 0, errors.Wrap(err, "last id from software")
+	}
+	return uint(id), nil
+}
+
+func (d *Datastore) insertNewInstalledHostSoftware(
+	tx *sqlx.Tx,
+	hostID uint,
+	currentIdmap map[string]uint,
+	incomingBitmap map[string]bool,
+) error {
+	var insertsHostSoftware []interface{}
```

Comment:
```suggestion
	var insertsHostSoftware []uint
```

Reply:
I agree with it, but we can't do that because otherwise later on when we use it we would have to cast it to []interface{}, which involves going through each item and casting them individually.